### PR TITLE
WIP: align Qwen3.5 PaddleFleet numerics

### PR DESCRIPTION
**Pull request by zrr1999**
## Status
Work-in-progress Qwen3.5 raw bit-exact alignment for CUDA/BF16 world4 EP4.

Current formal text loss remains 0/10 exact. EXP-021 localizes the native GDN first-bad; EXP-023 OFFx3 is baseline-identical and ON validation is still running.

## Scope
- accuracy-compatible GDN projection/parser/Conv/L2/recurrence candidates
- eager attention and vision alignment paths
- deterministic/no-extension support and focused tests

## Validation
- PaddleFleet Qwen3.5 focused tests: 16 passed
- Ruff E9: passed
- formal Paddle 10-step transaction: runs successfully

Not ready for merge; candidate retention, full forward/loss, backward, optimizer, checkpoint, and multimodal gates remain open.

### CODE REVIEW

File: packages/paddlefleet_ops/src/paddlefleet_ops/__init__.py [hunk 1]
```diff
@@ -238,8 +238,14 @@ def _sonic_moe_requirement(
 _FLASH_MASK_AVAILABLE = False
 _CUDNN_FRONTEND_AVAILABLE = False
 _FAST_HADAMARD_TRANSFORM_AVAILABLE = False
-
-if paddle.is_compiled_with_cuda():
+_NO_EXT = os.environ.get("PADDLEFLEET_OPS_NO_EXT", "0").lower() in {
```
Comment (risemeup1111) [inline packages/paddlefleet_ops/src/paddlefleet_ops/__init__.py:241]: P1 no-ext 构建状态没有随 wheel 持久化。`setup_install_no_extension()` 不会生成 `.ops` 或准备 ecosystem 库，但安装后的新进程会在这里重新读取环境变量；若用户只在构建命令中设置 `PADDLEFLEET_OPS_NO_EXT=1`，运行时未继续导出它，CUDA 环境会把这些能力重新标为可用并进入后续 ecosystem 导入，no-ext wheel 因缺少对应模块而无法正常导入。

请在构建产物中写入可查询的 no-ext 标记，或按实际模块是否存在来决定可用性，并补充“no-ext 构建 wheel -> 清除构建环境变量 -> CUDA 进程导入”的回归测试。

File: src/paddlefleet/transformer/gated_delta_net.py [hunk 2]
```diff
@@ -270,32 +567,48 @@ def _build_padding_mask(
...
+                valid_tokens = attention_mask
+                full_seq = attention_mask.shape[-1]
+            elif (
+                attention_mask.shape[1] == 1
+                and attention_mask.shape[-2] == attention_mask.shape[-1]
+            ):
+                # PaddleFormers dense masks use 1.0 for an allowed QK edge and
+                # 0.0 for a masked edge. A valid causal query always retains its
+                # diagonal edge, while padding rows have a zero diagonal.
+                valid_tokens = paddle.diagonal(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/gated_delta_net.py:581]: P1 这里没有区分 4-D mask 的语义。仓库标准 bool mask 由 `attention_mask_func` 以 `True=masked-out` 使用，因此有效 causal token 的对角线是 `False`；直接取 diagonal 后会把所有有效 token 当成 0，随后在 `forward()` 中把 hidden states 全部清零。当前测试只覆盖了 PaddleFormers 的 float32 `1=allowed` 格式，未覆盖这条既有 bool 契约。

请按 dtype/格式显式处理：对标准 bool mask 反转对角线，或在无法可靠判别时明确拒绝；同时增加 4-D bool causal/padding mask 的回归测试，避免静默产生错误输出。

File: src/paddlefleet/transformer/transformer_layer.py [hunk 3]
```diff
@@ -592,6 +590,40 @@ def forward(
...
                 main_seq_len = hidden_states.shape[1]
+
+            # The non-flexible MTP collator pads input_ids and its dense causal
+            # mask with one lookahead position per MTP layer. Embedding converts
+            # those lookahead positions into separate MTP hidden-state chunks,
+            # so every decoder and MTP transformer block operates on
+            # main_seq_len tokens. Keep the dense mask on that same model-facing
+            # sequence; passing the padded mask onward either leaks a stale key
+            # column into full attention or makes GatedDeltaNet reject it.
+            attention_mask = dict_args.get("attention_mask", None)
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_layer.py:601]: P1 这段裁剪只覆盖基类 `TransformerLayer.forward()`。`HySparseTransformerLayer` 覆盖了 `forward()`，并通过 `_mtp_split()` 单独镜像 MTP 拆分，但该 helper 没有同步裁剪 `attention_mask`；`enable_hy_sparse_attention + MTP` 时 hidden sequence 已变为 `main_seq_len`，dense mask 仍是 `main_seq_len + num_nextn`，随后进入 attention 会发生 shape 不匹配或保留过期 key 列。

请把同一裁剪/校验逻辑同步到 `_mtp_split()`（或抽成两条路径共用的 helper），并为 HySparse + MTP 的 2-D/4-D mask 增加回归测试。

File: src/paddlefleet/transformer/gated_delta_net.py [hunk 4]
```diff
@@ -480,6 +856,9 @@ def forward(
                 initial_state=None,
                 output_final_state=False,
                 use_qk_l2norm_in_kernel=False,
+                use_accuracy_compatible_cumsum=(
+                    self.accuracy_compatible_gdn_recurrence
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/gated_delta_net.py:860]: P1 这组开关组合改变了训练主路径中的投影顺序、Conv1D、Q/K 归一化和 recurrence scan，但新增测试只分别复算 helper 输出及检查非法开关依赖；没有任何测试以这些开关为正值构造完整 `GatedDeltaNet` 并执行 backward。当前 PR 描述也明确说明 backward gate 尚未完成，因此参数梯度错误目前无法被回归测试发现。

请增加 BF16/CUDA 的端到端对齐用例：在完整开关组合下比较参考实现的 forward，并至少校验 input、in-proj、conv、`dt_bias`、`A_log`、norm 和 out-proj 的梯度；recurrence 开关应实际经过本行而不是直接测试 scan helper。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [no-ext wheel 状态持久化](https://github.com/PaddlePaddle/PaddleFleet/pull/1646#discussion_r3718997584) | P1 | 🚧 |
| 2 | [GDN 4-D bool mask 语义](https://github.com/PaddlePaddle/PaddleFleet/pull/1646#discussion_r3718999201) | P1 | 🚧 |
| 3 | [HySparse MTP mask 裁剪](https://github.com/PaddlePaddle/PaddleFleet/pull/1646#discussion_r3719001018) | P1 | 🚧 |
| 4 | [GDN BF16/CUDA 端到端反向覆盖](https://github.com/PaddlePaddle/PaddleFleet/pull/1646#discussion_r3719003266) | P1 | 🚧 |
| 5 | [PR 标题与描述](https://github.com/PaddlePaddle/PaddleFleet/pull/1646) | P3 | 🚧 |

PR 标题缺少规范 Tag，描述也缺少模板要求的三个精确章节。可直接替换为：

**标题**
```text
[Improvements] Align Qwen3.5 PaddleFleet numerics
```

**描述**
```markdown
### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library  | Environment Adaptation ] -->
Distributed Strategy
### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->
Improvements
### Description
<!-- Describe what you’ve done -->
- Add opt-in Qwen3.5 GDN numerical-alignment candidates for projection parsing, Conv1D, Q/K L2 normalization, and recurrence.
- Add eager packed vision attention and vision projection/interpolation alignment paths.
- Add deterministic no-extension installation support and focused tests.
- Trim dense attention masks after non-flexible MTP inputs are split.
- Current author-reported validation: 16 focused Qwen3.5 tests, Ruff E9, and a formal Paddle 10-step transaction.
- Remaining gates before merge: candidate retention, full forward/loss, BF16/CUDA backward, optimizer, checkpoint, and multimodal validation.
```

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
